Mark ScrollingNodes as dirty as optimization for scrolling tree commit

nmoucht opened the pull request:
#### 358ba2e1aa61dd81e74a781ffe044bf21add45e3
<pre>
Mark ScrollingNodes as dirty as optimization for scrolling tree commit
<a href="https://bugs.webkit.org/show_bug.cgi?id=256655">https://bugs.webkit.org/show_bug.cgi?id=256655</a>
rdar://109218372

Reviewed by NOBODY (OOPS!).

When a property on a ScrollingNode is changed, iterate up the ancestor
chain and mark each node as having a child with a changed property. When
committing the scrolling tree, ignore the branches with a node without
having a child with an update. The previous commit code used an unvisited
nodes vector to know when a node was removed. Since this requires visiting
every node, replace this with a vector on the scrolling state tree that tracks
nodes removed since the last commit. Remove use of univsited nodes list in favor
of using this new list of removed nodes. ScrollingStateNodes also had a property
m_scrolledSinceLastCommit. Since not every node is visited anymore, create a new
map on ScrollingTree m_nodesScrolledSinceLastCommit, which is added to after each
node is scrolled and cleared after each commit. There is still work needed for
OverflowScrollProxyNodes, as the maps containing active proxy nodes and their
related nodes are invalidated after each commit, and expect that we recreate them
as we iterate through the tree. We need a similar fix for positioned nodes as well.

* Source/WebCore/page/scrolling/ScrollingStateNode.cpp:
(WebCore::ScrollingStateNode::ScrollingStateNode):
(WebCore::ScrollingStateNode::setPropertyChanged):
(WebCore::ScrollingStateNode::setDirtyBitForAncestorChain):
(WebCore::ScrollingStateNode::dumpProperties const):
* Source/WebCore/page/scrolling/ScrollingStateNode.h:
(WebCore::ScrollingStateNode::resetChangedProperties):
(WebCore::ScrollingStateNode::setDirty):
(WebCore::ScrollingStateNode::hasDirtyChildren const):
* Source/WebCore/page/scrolling/ScrollingTree.cpp:
(WebCore::ScrollingTree::commitTreeState):
(WebCore::ScrollingTree::updateTreeFromStateNodeRecursive):
* Source/WebKit/Shared/RemoteLayerTree/RemoteScrollingCoordinatorTransaction.cpp:
(ArgumentCoder&lt;ScrollingStateScrollingNode&gt;::encode):
(ArgumentCoder&lt;ScrollingStateScrollingNode&gt;::decode):
</pre><!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/358ba2e1aa61dd81e74a781ffe044bf21add45e3

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/10104 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/10348 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/10601 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/11755 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/9788 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/10113 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/12338 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/10300 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/12707 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/10259 "Passed tests") | [❌ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/11027 "2 new passes 8 flakes 6 failures") | [❌ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/8544 "55 api tests failed or timed out") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/12140 "Built successfully") | 
| | [❌ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/8332 "8 flakes 3 failures") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/9155 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/16470 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/9433 "Passed tests") | [❌ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/9306 "3 failures") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/12567 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/9790 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/7952 "3 failures") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/8949 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/13187 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/9591 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

smfr commented on Source/WebCore/page/scrolling/ScrollingStateNode.h:
```diff
@@ -348,6 +351,8 @@ class ScrollingStateNode : public ThreadSafeRefCountedAndCanMakeThreadSafeWeakPt
     ThreadSafeWeakPtr<ScrollingStateNode> m_parent;
     std::unique_ptr<Vector<RefPtr<ScrollingStateNode>>> m_children;
 
+    bool m_childrenAreDirty { false };
```
I wonder if this can just be another bit in ScrollingStateNodeProperty ?

nmoucht committed:
Mark ScrollingNodes as dirty as optimization for scrolling tree commit
https://bugs.webkit.org/show_bug.cgi?id=256655
rdar://109218372

Reviewed by NOBODY (OOPS!).

When a property on a ScrollingNode is changed, iterate up the ancestor
chain and mark each node as having a child with a changed property. When
committing the scrolling tree, ignore the branches with a node without
having a child with an update. The previous commit code used an unvisited
nodes vector to know when a node was removed. Since this requires visiting
every node, replace this with a vector on the scrolling state tree that tracks
nodes removed since the last commit. Remove use of univsited nodes list in favor
of using this new list of removed nodes. ScrollingStateNodes also had a property
m_scrolledSinceLastCommit. Since not every node is visited anymore, create a new
map on ScrollingTree m_nodesScrolledSinceLastCommit, which is added to after each
node is scrolled and cleared after each commit. There is still work needed for
OverflowScrollProxyNodes, as the maps containing active proxy nodes and their
related nodes are invalidated after each commit, and expect that we recreate them
as we iterate through the tree. We need a similar fix for positioned nodes as well.

* Source/WebCore/page/scrolling/ScrollingStateNode.cpp:
(WebCore::ScrollingStateNode::ScrollingStateNode):
(WebCore::ScrollingStateNode::setPropertyChanged):
(WebCore::ScrollingStateNode::setDirtyBitForAncestorChain):
(WebCore::ScrollingStateNode::dumpProperties const):
* Source/WebCore/page/scrolling/ScrollingStateNode.h:
(WebCore::ScrollingStateNode::resetChangedProperties):
(WebCore::ScrollingStateNode::setDirty):
(WebCore::ScrollingStateNode::hasDirtyChildren const):
* Source/WebCore/page/scrolling/ScrollingTree.cpp:
(WebCore::ScrollingTree::commitTreeState):
(WebCore::ScrollingTree::updateTreeFromStateNodeRecursive):
* Source/WebKit/Shared/RemoteLayerTree/RemoteScrollingCoordinatorTransaction.cpp:
(ArgumentCoder<ScrollingStateScrollingNode>::encode):
(ArgumentCoder<ScrollingStateScrollingNode>::decode):
